core/src/main/scala/org/bykn/bosatsu/TotalityCheck.scala

```diff
@@ -287,126 +449,127 @@ case class TotalityCheck(inEnv: TypeEnv[Any]) {
+                  // a mismatch
+                  unifyUnion(getProd(la).difference(lp, rp)
+                    .map(PositionalStruct(ln, _): Pattern[Cons, Type]))
                 }
+                else (left :: Nil)
+              case (PositionalStruct(n, as), rp@ListPat(_)) =>
+                structToList(n, as) match {
+                  case Some(lp) => difference(lp, rp)
+                  case None => left :: Nil
+                }
+              case (lp@ListPat(_), PositionalStruct(n, as)) =>
+                structToList(n, as) match {
+                  case Some(rp) => difference(lp, rp)
+                  case None => left :: Nil
+                }
+              case (_, PositionalStruct(nm, _)) if isTop(left) =>
+                inEnv.definedTypeFor(nm) match {
+                  case None =>
+                    // just assume this is infinitely bigger than unknown
+                    // types
+                    left :: Nil
+                  case Some(dt) =>
+                    dt.constructors.flatMap {
+                      case cf if (dt.packageName, cf.name) == nm =>
+                        // we can replace _ with Struct(_, _...)
+                        val newWild = PositionalStruct(nm, cf.args.map(_ => WildCard))
+                        difference(newWild, right)
+
+                      case cf =>
+                        // TODO, this could be smarter
+                        // we need to learn how to deal with typed generics
+                        def argToPat[A](t: (A, Type)): Pattern[Cons, Type] =
+                          if (Type.hasNoVars(t._2)) Annotation(WildCard, t._2)
+                          else WildCard
+
+                        PositionalStruct((dt.packageName, cf.name), cf.args.map(argToPat)) :: Nil
+                    }
+                }
+              case (_, _) =>
+                // ill-typed
+                if (isTop(left)) {
+                  right match {
+                    case StrPat(_) => difference(StrPat.Wild, right)
+                    case ListPat(_) => difference(ListPat.Wild, right)
+                    // we already checked for right as PositionalStruct
+                    // we can't compute top - literal
+                    case _ =>
+                      // we can't solve this
+                      left :: Nil
+                  }
+                }
+                else left :: Nil
             }
-          case (_, _) =>
-            // ill-typed
-            if (isTop(right)) Nil
-            else if (isTop(left)) {
-              right match {
-                case StrPat(_) => difference(StrPat.Wild, right)
-                case ListPat(_) => difference(ListPat.Wild, right)
-                case _ =>
-                  // we can't solve this
-                  left :: Nil
-              }
-            }
-            else left :: Nil
-        }
+          }
 
       def isTop(p: Pattern[Cons, Type]): Boolean =
```

Comment:
Can we make this as simple as checking `relate(Wildcard, p) == Rel.Same`

Reply:
no, we can't. `relate` with `Wildcard` always results in the `solve` branch happening, which calls unifyUnion, which needs isTop.